### DISCUSSION TITLE
Converting an 8-bit PNG with a palette to another palette using a specified GIMP palette file

### DISCUSSION THREAD

**Discussion by bigyihsuan** [ImageMagick/ImageMagick#7023] [Help]
### ImageMagick version

7.1.1-27 (Beta)

### Operating system, version and so on

Ubuntu 22.04.3 LTS on Windows 10 x86_64 (WSL)

### Description

I'm working on a NewGRF for [OpenTTD](https://github.com/OpenTTD/OpenTTD) have over 200 individual 8-bit PNG spritesheets with palettes, and want to use ImageMagick to convert them en masse to a different palette.

Below is an example of a spritesheet I'm trying to convert:

![train_28](https://github.com/ImageMagick/ImageMagick/assets/12767476/ab7909eb-549e-4c22-838b-3019a3e47935)

with `$ identify train_28.png` saying that it is a `train_28.png PNG 253x166 253x166+0+0 8-bit sRGB 256c 3994B 0.000u 0:00.002`.

However, the palette files I have are GIMP GPL files (linked below, renamed to TXT. These are just text files, it seems, mapping an index to an RGB value).
I can manually do the conversion in GIMP by doing

1. Image > Mode > RGB, then
2. Image > Mode > Indexed > Use custom palette, and select the new palette.

Usually this process works. However, there are over 200 images to do this process with, which is why I'm asking here about ImageMagick.

I've found [this link](https://www.imagemagick.org/Usage/quantize/#remap) about the `-remap` option but that seems to take in another image to take the palette from, rather than an actual palette.

Is there a way of doing this in ImageMagick?

[ttd-newgrf-dos.txt](https://github.com/ImageMagick/ImageMagick/files/13920373/ttd-newgrf-dos.txt)
[ttd-newgrf-win.txt](https://github.com/ImageMagick/ImageMagick/files/13920374/ttd-newgrf-win.txt)

**Comment by fmw42** [comment]
My suggestion would be to write a script loop to convert each color in the color table to 1 pixel images and append into one image. Then use that with -remap.  I do not know that Imagemagick can swap color tables.  Perhaps someone else has a better idea or knows how to do that.

**Reply by fmw42** [reply]
Perhaps a better way/concept is to use -clut (see https://imagemagick.org/Usage/color_mods/#clut).  You would convert your new/desired colormap into 3 "gradient" images in the order listed for the index, one gradient for each channel in the color map corresponding to each channel in the image.  Then use -clut 3 time, once for each channel in the image (after separating channels) using the corresponding colormap image. Then combine the channels together.

**Reply by fmw42** [reply]
Sorry, this is not correct. Ignore this message above.

**Comment by snibgo** [comment]
ImageMagick does not contain CLI options to directly manipulate palettes. If you want to create or modify an image to contain a palette of specified colours, in a specified order, IM doesn't have options to do that.

**Reply by fmw42** [reply]
Which of your two files is the existing colormap and which is the desired colormap?

**Reply by fmw42** [reply]
Why are you using the beta release? It may not be final.

**Comment by fmw42** [comment]
You could script it in Python PIL.  See https://stackoverflow.com/questions/62898521/how-do-i-use-pillows-image-convert-with-a-custom-color-palette
